### PR TITLE
Make Swift result variables persistent

### PR DESCRIPTION
**Pull request by adrian-prantl**
While also being more efficient and shorter, this is done mainly in preparation for followup changes.

### CODE REVIEW

**Comment by adrian-prantl** [comment]
rdar://113352990

**Comment by adrian-prantl** [comment]
@swift-ci test

**Review by kastiglione — COMMENTED** [review]
Why was this a multimap, and why can it now be a (single) map?

I did a quick read of the places where `equal_range` was used, but on first look it's not clear to me why it was a multimap, nor is it clear what `DeclsAreEquivalent` was used for.

Note: `DeclsAreEquivalent` is no longer used now as of this change.

**Comment by adrian-prantl** [comment]
> Why was this a multimap, and why can it now be a (single) map?
> 

I converted a `multimap<String, Decl *>` into a `Stringmap<SmallVector<Decl *>>`

> I did a quick read of the places where `equal_range` was used, but on first look it's not clear to me why it was a multimap, nor is it clear what `DeclsAreEquivalent` was used for.
> 
> Note: `DeclsAreEquivalent` is no longer used now as of this change.

Let me check ...

**Comment by adrian-prantl** [comment]
Fix call to the wrong version of SmallVector.erase()

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
Added a second NFC refactoring on top of this patch.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
The last patch in the series is still WIP, in the sense of that I still need to add a test for it.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test windows

**Comment by adrian-prantl** [comment]
@augusto2112 @kastiglione Would you mind taking another look at this? The goal here is to make Swift result variables portable between different SwiftASTContext instances.

File: lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp [hunk 1]
```diff
@@ -401,11 +399,10 @@ class LLDBREPLNameLookup : public LLDBNameLookup {
     if (NameStr.empty())
       return false;
 
-    ConstString name_const_str(NameStr);
-    if (m_log) {
-      m_log->Printf("[LLDBREPLNameLookup::lookupAdditions (%u)] Searching for %s",
-                    count, name_const_str.AsCString("<anonymous>"));
-    }
+    if (m_log)
+      m_log->Printf(
```
Comment (augusto2112) [inline lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp]: Nit: you could use `LLDB_LOGF` here

**Comment by adrian-prantl** [comment]
ping :-)

**Review by augusto2112 — APPROVED** [review]

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h [hunk 2]
```diff
@@ -1037,8 +1014,45 @@ class SwiftASTContextForExpressions : public SwiftASTContext {
...
+  bool CacheUserImports(lldb::ProcessSP process_sp,
+                        swift::SourceFile &source_file, Status &error);
+
+protected:
+  /// These are the names of modules that we have loaded by hand into
+  /// the Contexts we make for parsing.
+  HandLoadedModuleSet m_hand_loaded_modules;
+
+  /// This returns the list of hand-loaded modules.
+  HandLoadedModuleSet GetHandLoadedModules() { return m_hand_loaded_modules; }
```
Comment (kastiglione) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h]: Could this be returned as a const reference instead of by copy?

```suggestion
  const HandLoadedModuleSet &GetHandLoadedModules() { return m_hand_loaded_modules; }
```

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h [hunk 3]
```diff
@@ -1037,8 +1014,45 @@ class SwiftASTContextForExpressions : public SwiftASTContext {
 
   void ModulesDidLoad(ModuleList &module_list);
 
-private:
-  std::unique_ptr<SwiftPersistentExpressionState> m_persistent_state_up;
+  typedef llvm::StringMap<swift::AttributedImport<swift::ImportedModule>>
+      HandLoadedModuleSet;
+
+  // This just adds this module to the list of hand-loaded modules, it doesn't
+  // actually load it.
```
Comment (kastiglione) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h]: It doesn't only _add_ the list of modules, it could overwrites an existing value, if any.

File: lldb/source/Plugins/TypeSystem/Swift/StoringDiagnosticConsumer.h [hunk 4]
```diff
@@ -15,6 +15,8 @@
 
 #include "Plugins/ExpressionParser/Swift/SwiftDiagnostic.h"
 
+#include "lldb/Utility/StreamString.h"
```
Comment (kastiglione) [inline lldb/source/Plugins/TypeSystem/Swift/StoringDiagnosticConsumer.h:18]: minor: seems unrelated to this PR

File: lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp [hunk 5]
```diff
@@ -248,17 +253,16 @@ class LLDBExprNameLookup : public LLDBNameLookup {
...
 
-    ConstString name_const_str(NameStr);
-    if (m_log) {
-      m_log->Printf("[LLDBExprNameLookup::lookupAdditions (%u)] Searching for %s",
-                    count, name_const_str.AsCString("<anonymous>"));
-    }
+    if (m_log)
+      m_log->Printf(
+          "[LLDBExprNameLookup::lookupAdditions (%u)] Searching for %s", count,
+          NameStr.empty() ? "<anonymous>" : NameStr.str().c_str());
```
Comment (kastiglione) [inline lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp]: `NameStr` can't be empty at this point, because of the early return guard clause just prior.

File: lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp [hunk 6]
```diff
@@ -401,34 +409,34 @@ class LLDBREPLNameLookup : public LLDBNameLookup {
...
 
-    ConstString name_const_str(NameStr);
-    if (m_log) {
-      m_log->Printf("[LLDBREPLNameLookup::lookupAdditions (%u)] Searching for %s",
-                    count, name_const_str.AsCString("<anonymous>"));
-    }
+    if (m_log)
+      m_log->Printf(
+          "[LLDBREPLNameLookup::lookupAdditions (%u)] Searching for %s", count,
+          NameStr.empty() ? "<anonymous>" : NameStr.str().c_str());
```
Comment (kastiglione) [inline lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp]: same here, `NameStr` will not be empty.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test windows

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test windows

**Comment by adrian-prantl** [comment]
@swift-ci test windows
